### PR TITLE
feat: shadow prover support for RPC followers

### PR DESCRIPTION
**Pull request by rkrasiuk**
## Summary

Enables RPC-only follower nodes to run the detached shadow prover without requiring sequencer or settlement keys. The follower observes finalized submitBatch transactions on L1, decodes the accepted quorum-certificate inputs, derives the exact anchor used by settlement, and waits for the corresponding Zone block range to become canonical locally before enqueueing the proof.

### CODE REVIEW

**Comment by rkrasiuk** [comment]
cyclops audit fast

File: crates/node/src/shadow_prover.rs [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+    }
+
+    async fn fetch_submit_batch_calls(
+        &self,
+        tx_hash: B256,
+    ) -> Result<Vec<ZonePortal::submitBatchCall>> {
+        let tx = self
+            .l1_provider
+            .client()
+            .request::<_, serde_json::Value>("eth_getTransactionByHash", (tx_hash,))
```
Comment (shekhirin) [inline crates/node/src/shadow_prover.rs]: can we use alloy provider method here?

File: crates/node/src/shadow_prover.rs [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+        let this = self.clone();
+        tokio::spawn(async move {
+            loop {
+                if recovery_sender.is_closed() {
+                    return;
+                }
+                match this.recover_recent_submissions().await {
+                    Ok(recovered) => {
+                        for submission in recovered {
+                            if recovery_sender.send(submission).is_err() {
```
Comment (shekhirin) [inline crates/node/src/shadow_prover.rs]: can we send to this channel directly from `recover_recent_submissions`, so that we don't accumulate them all in memory first

File: crates/node/src/shadow_prover.rs [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+        anchor: ShadowProofAnchor,
+    ) -> Result<()> {
+        let from = loop {
+            if batch.prev_block_hash.is_zero() {
+                break 1;
+            }
+            if let Some(previous) = self.zone_provider.block_number(batch.prev_block_hash)? {
+                break previous.saturating_add(1);
+            }
+            tokio::time::sleep(POLL_INTERVAL).await;
```
Comment (shekhirin) [inline crates/node/src/shadow_prover.rs:268]: do we expect this to be hit due to a race between new block processing and proving?

Comment (rkrasiuk) [reply crates/node/src/shadow_prover.rs:268]: not exactly with proving, but rather the local state of the follower node. it might lag behind the quorum participating zone nodes

File: crates/node/src/shadow_prover.rs [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+            }
+        });
+    }
+
+    async fn process_submission(&self, submission: &FinalizedBatchSubmission) -> Result<()> {
+        let call = self
+            .fetch_submit_batch_calls(submission.transaction_hash)
+            .await?
+            .into_iter()
+            .find(|call| call_matches_event(call, &submission.event))
```
[the remark this answers is not in the document]
Comment (shekhirin) [reply crates/node/src/shadow_prover.rs:107]: not a concern, batch submission txs always have one `submitBatch` call

**Review by shekhirin — DISMISSED** [review]

File: crates/l1/src/subscriber.rs [hunk 5]
```diff
@@ -379,6 +394,9 @@ pub struct L1SubscriberConfig {
     /// Optional sink that receives leadership transitions before the block that
     /// recorded them is enqueued.
     pub leadership_sink: Option<Arc<dyn LeadershipSink>>,
+    /// Optional observational channel for finalized accepted batch submissions.
+    pub finalized_batch_submissions:
+        Option<tokio::sync::mpsc::UnboundedSender<FinalizedBatchSubmission>>,
```
Comment (0xKitsune) [inline crates/l1/src/subscriber.rs]: Not blocking but lets plan to move this to the `L1Subscriber` struct pending https://github.com/tempoxyz/zones/pull/1356 landing.

File: crates/node/src/shadow_prover.rs [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+
+        info!(
+            target: "zone::node::shadow_prover",
+            zone_from = from,
+            zone_to = to,
+            anchor_number = anchor.number,
+            anchor_hash = %anchor.hash,
+            "Queueing finalized quorum-certified batch for shadow proving"
+        );
+        self.prover
```
Comment (0xKitsune) [inline crates/node/src/shadow_prover.rs]: It looks like a batch that does not end in `finalizeWithdrawalBatch` can still pass here because the SPF accepts open snapshots and copies the expected batch index.

File: crates/l1/src/subscriber.rs [hunk 7]
```diff
@@ -903,6 +933,34 @@ impl L1Subscriber {
...
+                        && log.topic0() == Some(&ZonePortal::BatchSubmitted::SIGNATURE_HASH)
+                    {
+                        match (
+                            ZonePortal::BatchSubmitted::decode_log(&log.inner),
+                            log.log_index,
+                        ) {
+                            (Ok(event), Some(log_index)) => {
+                                finalized_batches.push(FinalizedBatchSubmission {
+                                    block_number,
+                                    transaction_hash: receipt.transaction_hash(),
```
Comment (0xKitsune) [inline crates/l1/src/subscriber.rs]: Does the fetched transaction needs to be authenticated against the finalized block since receipt root verification does not cover `transaction_hash` or `log_index` RPC metadata.

File: crates/node/src/shadow_prover.rs [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+            }
+        });
+    }
+
+    async fn process_submission(&self, submission: &FinalizedBatchSubmission) -> Result<()> {
+        let call = self
+            .fetch_submit_batch_calls(submission.transaction_hash)
+            .await?
+            .into_iter()
+            .find(|call| call_matches_event(call, &submission.event))
```
Comment (0xKitsune) [inline crates/node/src/shadow_prover.rs:107]: Maybe a nit but IIUC a multicall tx can map multiple logs to the first matching `submitBatch` call because call order and the log index are ignored.

File: crates/node/src/shadow_prover.rs [hunk 9]
```diff
@@ -0,0 +1,360 @@
...
+        mut submissions: UnboundedReceiver<FinalizedBatchSubmission>,
+        recovery_sender: UnboundedSender<FinalizedBatchSubmission>,
+    ) {
+        self.spawn_recovery(recovery_sender);
+
+        let mut observed = HashSet::new();
+        while let Some(submission) = submissions.recv().await {
+            let observation_id = (submission.transaction_hash, submission.log_index);
+            if observed.insert(observation_id) {
+                self.spawn_submission_retry(submission);
```
Comment (0xKitsune) [inline crates/node/src/shadow_prover.rs]: It looks like this can bypass the queue bound because every observation is retained forever and gets a task that can also wait forever.

File: crates/node/src/shadow_prover.rs [hunk 10]
```diff
@@ -0,0 +1,360 @@
...
+        return Ok(tempo_block_number);
+    }
+    ensure!(
+        recent_tempo_block_number > tempo_block_number,
+        "submitted ancestry anchor does not follow its Tempo checkpoint"
+    );
+    Ok(recent_tempo_block_number)
+}
+
+#[cfg(test)]
```
Comment (0xKitsune) [inline crates/node/src/shadow_prover.rs:351]: Looks like we could also use tests covering recovery, calldata binding and rejection of non boundary submissions.

File: crates/l1/src/subscriber.rs [hunk 11]
```diff
@@ -828,6 +846,17 @@ impl L1Subscriber {
                     .block_tracker
                     .record_with_portal_events(anchor, events.clone())?;
             }
+            if let Some(sender) = &self.config.finalized_batch_submissions {
+                for submission in finalized_batches {
+                    if sender.send(submission).is_err() {
```
Comment (0xKitsune) [inline crates/l1/src/subscriber.rs]: It looks like a failed send means the observer is permanently dead, but we only warn and silently drop every later submission.

File: crates/l1/src/tests.rs [hunk 12]
```diff
@@ -1652,6 +1653,87 @@ fn corrupt_recognized_portal_log(portal: Address) -> Log {
...
+    assert_eq!(submissions.len(), 1);
+    assert_eq!(submissions[0].block_number, 10);
+    assert_eq!(submissions[0].transaction_hash, B256::with_last_byte(0xaa));
+    assert_eq!(submissions[0].log_index, 4);
+    assert_eq!(submissions[0].event.nextBlockHash, B256::repeat_byte(0x22));
+}
+
+#[test]
+fn finalized_batch_observer_errors_do_not_fence_ingestion() {
+    use alloy_sol_types::SolEvent as _;
```
Comment (0xKitsune) [inline crates/l1/src/tests.rs]: nit: can we import this at the top of the tests mod once rather than in each test function?

File: crates/node/src/node.rs [hunk 13]
```diff
@@ -198,6 +199,17 @@ pub struct ZoneSequencerAddOnsConfig {
...
 
+/// Configuration for detached shadow proving that does not require sequencer keys.
+#[derive(Debug, Clone)]
+pub struct ZoneShadowProverAddOnsConfig {
+    /// Zone ID bound into SPF public inputs.
+    pub zone_id: u32,
+    /// EIP-2935 history settings used to recover recently finalized submissions.
+    pub batch_anchor_config: BatchAnchorConfig,
+    /// Remote prover TCP address. When absent, execute the SPF in-process.
+    pub prover_address: Option<String>,
```
Comment (0xKitsune) [inline crates/node/src/node.rs]: minor nit: might be nice to make this an enum. Something like:

```rust
pub enum ProverRuntime {
    InProcess,
    Remote(IPAddr)
}
```

Feel free to ignore this if preferred.

File: crates/l1/src/subscriber.rs (hunk 5 above)
Comment (rkrasiuk) [reply crates/l1/src/subscriber.rs]: reasonable, let's do in the follow up

File: crates/node/src/shadow_prover.rs [hunk 14]
```diff
@@ -0,0 +1,386 @@
...
+
+        info!(
+            target: "zone::node::shadow_prover",
+            zone_from = from,
+            zone_to = to,
+            anchor_number = anchor.number,
+            anchor_hash = %anchor.hash,
+            "Queueing finalized quorum-certified batch for shadow proving"
+        );
+        self.prover
```
[the remark this answers is not in the document]
Comment (rkrasiuk) [reply crates/node/src/shadow_prover.rs]: fixed in https://github.com/tempoxyz/zones/pull/1324/changes/7a9dd64b10da3dbd6cb287feea5884baace9806b, finalized-settlement jobs now call `validate_settlement_boundary` before proving.

File: crates/node/src/shadow_prover.rs (hunk 6 above)
Comment (rkrasiuk) [reply crates/node/src/shadow_prover.rs]: addressed in https://github.com/tempoxyz/zones/pull/1324/changes/7a9dd64b10da3dbd6cb287feea5884baace9806b

File: crates/l1/src/subscriber.rs (hunk 7 above)
Comment (rkrasiuk) [reply crates/l1/src/subscriber.rs]: addressed in https://github.com/tempoxyz/zones/pull/1324/changes/adf2005b50487d2edbfeab322d3a867c09cc043f

File: crates/l1/src/subscriber.rs (hunk 11 above)
Comment (rkrasiuk) [reply crates/l1/src/subscriber.rs]: addressed in https://github.com/tempoxyz/zones/pull/1324/commits/d9c2b9752adfd94ef625894a3635554affefbc2a

File: crates/l1/src/subscriber.rs [hunk 15]
```diff
@@ -36,6 +37,54 @@ pub struct AuthenticatedPortalLogs {
...
+                    target: "zone::l1::subscriber",
+                    block_number = block.number,
+                    error = ?err,
+                    "Could not decode finalized batch submission for observer"
+                ),
+            }
+        }
+    }
+    submissions
+}
```
Comment (shekhirin) [inline crates/l1/src/subscriber.rs]: definitions are already way too out of order in this file, but can we move this function below where free functions are

**Comment by rkrasiuk** [comment]
cyclops audit fast

**Review by shekhirin — DISMISSED** [review]

**Review by shekhirin — DISMISSED** [review]

File: crates/node/src/shadow_prover.rs [hunk 16]
```diff
@@ -0,0 +1,400 @@
...
+                        error!(
+                            target: "zone::node::shadow_prover",
+                            l1_block_hash = %submission.block.hash,
+                            transaction_index = submission.transaction_index,
+                            log_index = submission.log_index,
+                            error = ?err,
+                            attempts = attempt,
+                            "Failed to process finalized batch submission; retry budget exhausted"
+                        );
+                        return;
```
Comment (0xKitsune) [inline crates/node/src/shadow_prover.rs:71]: It looks like retry exhaustion or a temporarily missing block permanently drops the finalized submission because there is no persisted cursor or startup replay.

File: crates/node/src/shadow_prover.rs [hunk 17]
```diff
@@ -0,0 +1,400 @@
...
+            anchor_number = anchor.number,
+            anchor_hash = %anchor.hash,
+            "Queueing finalized quorum-certified batch for shadow proving"
+        );
+        self.prover
+            .enqueue_with_anchor(from, to, batch, anchor)
+            .await
+    }
+
+    async fn wait_for_local_range(&self, to: u64, batch: &BatchData) -> Result<u64> {
```
Comment (0xKitsune) [inline crates/node/src/shadow_prover.rs:260]: nit: can we add some brief/succinct docs to these fns?

**Review by 0xKitsune — DISMISSED** [review]

File: crates/node/src/shadow_prover.rs (hunk 16 above)
Comment (rkrasiuk) [reply crates/node/src/shadow_prover.rs:71]: i intentionally removed replay for the shadow prover. i think we should get an alert on failure and investigate it if it failed beyond retries. adding state tracking specifically for this task is an overkill imo (which batches were validate or not)

**Comment by 0xKitsune** [comment]
cyclops audit fast

**Review by 0xKitsune — APPROVED** [review]
